### PR TITLE
Error on duplicate types and endpoint names

### PR DESCRIPTION
**Pull request by EdwinOuwehand**
## Description
This PR introduces validation to detect and report duplicate definitions (endpoints, types, channels) within Wirespec files. It ensures unique naming across the codebase, preventing ambiguity and improving the clarity of error reporting for developers. This enhancement prevents ambiguity in the Wirespec language and ensures consistent behavior across different compilation targets.

## Type of Change
- [x] Feature (new functionality)
- [ ] Bug fix (resolves an issue)
- [ ] Documentation update
- [ ] Code refactoring (no functional changes)
- [ ] Performance improvement
- [ ] Build/CI pipeline changes
- [ ] Other (please describe):

## Checklist
- [ ] I have followed the [contribution guidelines](https://github.com/flock-community/wirespec/blob/main/CONTRIBUTING.md)
- [x] I have written tests for my changes (as described in the PR details)
- [ ] I have written code in a functional style (using [Arrow](https://arrow-kt.io/) where appropriate)
- [ ] I have updated the documentation if necessary
- [x] My changes pass all tests (as described in the PR details)
- [ ] I have written a [conventional commit message](https://www.conventionalcommits.org/)

## Additional Context
**Implementation Details:**

* Added new validation error types:
    * `DuplicateEndpointError`: Reports duplicate endpoint names.
    * `DuplicateTypeError`: Reports duplicate type names.
    * `DuplicateChannelError`: Reports duplicate channel names.
* Introduced a `Validator` component responsible for checking duplicate definitions:
    * Performs validation across all modules.
    * Collects and reports all violations found, rather than stopping after the first one.
    * Integrated this validator into the existing parsing pipeline.
* Comprehensive unit tests were added to cover various scenarios:
    * Validation of unique definitions.
    * Detection of duplicates within the same file.
    * Detection of duplicates across different files.
    * Handling of multiple validation violations simultaneously.

## Breaking Changes
None.

### CODE REVIEW

File: src/compiler/core/src/commonTest/kotlin/community/flock/wirespec/compiler/core/validate/ValidatorTest.kt [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+           |type Foo { str: String }
+        """.trimMargin()
+
+        val source2 = """
+            |type Foo { str: String }
+        """.trimMargin()
+
+        validate(source1, source2)
+            .shouldBeRight()
+    }
```
Comment (ZzAve) [inline src/compiler/core/src/commonTest/kotlin/community/flock/wirespec/compiler/core/validate/ValidatorTest.kt:85]: I don't really understand the validity of this test, maybe in conjunction with the upcoming import functionality.

There is a possibility to process multiple ws files in one go, by specifying a folder, or a list of files (at least for the maven plugin, but I think this goes for the other tools as well). I think we also need to think about the output files. Cause if you define `type Foo` in two ws files, but process them at the same time (or one after the other), the resulting language files still have a conflict (if put in different files), or, potentially worse, overrides if the output files are the same

e.g. consider compiling  files `first.ws` and `second.ws` to kotlin in package `x`:

- `first.ws` --> this generates `Foo.kt` in package `x`
- `second.ws` --> this generates `Foo.kt` in package `x`, overwriting the already existing one

Do you think we can somehow ensure this kind of clashing / overwriting won't happen?

File: src/compiler/core/src/commonTest/kotlin/community/flock/wirespec/compiler/core/validate/ValidatorTest.kt [hunk 2]
```diff
@@ -0,0 +1,159 @@
+package community.flock.wirespec.compiler.core.validate
```
Comment (ZzAve) [inline src/compiler/core/src/commonTest/kotlin/community/flock/wirespec/compiler/core/validate/ValidatorTest.kt:1]: I'm curious to know how we want to deal with overlapping type vs endpoint declarations. IIUC, we split the generated code into files, which could mean that a type, a channel and an endpoint could potentially clash?


This example, although following a bad naming convention, is expected to give some issues, right? :
```
type SomeEndpoint{
  id: String
}

endpoint Some GET /Test1 -> {
  200 -> String
}
endpoint SomeEndpoint GET /Test1 -> {
  200 -> String
}

type SomeChannel {
  id: String
}

channel Some -> String
channel SomeChannel -> String
```

File: src/compiler/core/src/commonTest/kotlin/community/flock/wirespec/compiler/core/validate/ValidatorTest.kt [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+import community.flock.wirespec.compiler.core.validate
+import community.flock.wirespec.compiler.utils.NoLogger
+import io.kotest.assertions.arrow.core.shouldBeLeft
+import io.kotest.assertions.arrow.core.shouldBeRight
+import io.kotest.matchers.collections.shouldContainAll
+import io.kotest.matchers.shouldBe
+import io.kotest.matchers.types.shouldBeInstanceOf
+import kotlin.test.Test
+
+class ValidatorTest {
```
Comment (ZzAve) [inline src/compiler/core/src/commonTest/kotlin/community/flock/wirespec/compiler/core/validate/ValidatorTest.kt:22]: Great set of tests, covering a lot of the edge cases. Can we maybe add a happy scenario test as well, as to validate quickly that 'normal' usecase are not affected by changes we make to the validator?

File: src/compiler/core/src/commonMain/kotlin/community/flock/wirespec/compiler/core/exceptions/ValidationError.kt [hunk 4]
```diff
@@ -15,3 +15,15 @@ class EmptyModule :
...
+    ValidationError(
+        coordinates = Token.Coordinates(),
+        message = "Endpoint '$endpointName' is already defined",
+    )
+
+class DuplicateTypeError(typeName: String) :
+    ValidationError(
+        coordinates = Token.Coordinates(),
+        message = "Type '$typeName' is already defined",
+    )
```
Comment (ZzAve) [inline src/compiler/core/src/commonMain/kotlin/community/flock/wirespec/compiler/core/exceptions/ValidationError.kt:29]: Should we add DuplicateChannelError to this list as well, just to make it complete?

**Review by ZzAve — COMMENTED** [review]
This is a great feature to add! I'm sure @AngeloGroot is gonna love this as well, seeing the feature request originally came from him.

I got a few comments and pointers here and there. Very happy with the foundational work 💪

File: src/compiler/core/src/commonTest/kotlin/community/flock/wirespec/compiler/core/validate/ValidatorTest.kt (hunk 2 above)
Comment (EdwinOuwehand) [reply src/compiler/core/src/commonTest/kotlin/community/flock/wirespec/compiler/core/validate/ValidatorTest.kt:1]: This PR will solve exactly this problem 😄: 

https://github.com/flock-community/wirespec/pull/456

File: src/compiler/core/src/commonTest/kotlin/community/flock/wirespec/compiler/core/validate/ValidatorTest.kt (hunk 1 above)
Comment (EdwinOuwehand) [reply src/compiler/core/src/commonTest/kotlin/community/flock/wirespec/compiler/core/validate/ValidatorTest.kt:85]: So I agree that this is not ideal. I also don't think we should change anything at this time. 

Let's first establish why similar validation for endpoints is fine: it is because endpoints are in effect always a global definition and should be unique in that context. 

For types the scope is potentially narrower, though not exclusively local, it may still be shared by different endpoints/channels residing in different files. For now, this can be achieved by declaring the same type twice since we don't have imports yet. We'll then indeed write twice to `Foo.kt`. Apart from being ugly, this is problematic when the type definition isn't the same of course. Now in the future, this should be solved by imports indeed. However, we have not yet defined what should happen if you want to use the same type name in different places for different models. 

Assuming we don't want everything to simply be global, and we also don't yet want to break the current workaround, we are unfortunately stuck allowing such constructs.

File: src/compiler/core/src/commonTest/kotlin/community/flock/wirespec/compiler/core/validate/ValidatorTest.kt (hunk 3 above)
Comment (EdwinOuwehand) [reply src/compiler/core/src/commonTest/kotlin/community/flock/wirespec/compiler/core/validate/ValidatorTest.kt:22]: Done

File: src/compiler/core/src/commonTest/kotlin/community/flock/wirespec/compiler/core/validate/ValidatorTest.kt (hunk 1 above)
Comment (jerrevanveluw) [reply src/compiler/core/src/commonTest/kotlin/community/flock/wirespec/compiler/core/validate/ValidatorTest.kt:112]: True, but for consumers you would be able to namespace with different packages, so that point is mute

File: src/compiler/core/src/commonMain/kotlin/community/flock/wirespec/compiler/core/exceptions/ValidationError.kt (hunk 4 above)
Comment (EdwinOuwehand) [reply src/compiler/core/src/commonMain/kotlin/community/flock/wirespec/compiler/core/exceptions/ValidationError.kt:29]: Done

**Review by ZzAve — APPROVED** [review]
LGTM
